### DISCUSSION TITLE
Make search independent of word order and fuzzy

### DISCUSSION THREAD

**Discussion by jecorn** [mealie-recipes/mealie#2335] [Feature Request]
**Before submitting this feature request I have**
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
Recipe search is currently highly dependent on search phrasing and is non-fuzzy. For example, "pinto beans" and "beans pinto" are different searches, and "eggs over easy" is different than "egg over easy". See https://github.com/hay-kot/mealie/issues/2325

**Suggest A Solution**
The different function call names to Levenshtein distance and trigram search in sqlite vs Postgres make this a bit complicated via the sqlalchemy interface. @fleshgolem therefore suggested a two-stage fix:

- Split all search terms on whitespace and combine those to a single AND query, just so that it does not depend on ordering
- Implement actual fuzzy search to account for typos etc.

This is safe, sane and makes a lot of sense.

However, I also did some benchmarking with RapidFuzz (https://maxbachmann.github.io/RapidFuzz/) to test the crazy idea of just pulling relevant database fields from **all recipes** and then fuzzing against everything to figure out what matches. This would be database independent and could be easily pythonized inside the search function, but possibly slower. I was surprised to instead find out that RapidFuzz is blazingly fast and this is actually very fast with even a large dataset.

I started with a test set of 5 different recipes and made a "meta string" of the "name", "description", "tags" and "recipeIngredient:note" fields for each. Running **10'000 iterations** of a four-word search against these meta strings (e.g. simulating 5 separate 10'000 recipe databases) using even the slowest RapidFuzz algorithms (partial_token_set_ratio and partial_token_sort_ratio) takes on average **0.1 seconds per 10'000 recipe search**. These algorithms tokenise and variably sort or set-ize the strings to make them independent of word order and remove repeated words. The fuzzy matching was excellent (stop words should be removed). A one-word search is even faster, at average 0.05 seconds per 10'000 recipe search.

Breaking the search up into tiered search of various fields could be even faster, if real-life usage suggests that the meta-string approach is problematic.

**Additional Information**
- [X] If the feature is accepted I'm willing to submit a PR to provide this feature
Sort of. I have never developed inside docker and am only passingly familiar with the mealie codebase. But I could share a the inner code for a potential RapidFuzz-based change with someone who has more experience with the mealie code integration workflow (e.g. testing etc)

**Comment by michael-genson** [comment]
If you want to give it a shot, here's the dev getting started guide:
https://nightly.mealie.io/contributors/developers-guide/starting-dev-server/

And here's the core search logic: https://github.com/hay-kot/mealie/blob/de4debe74963ef48f057588c666ca2e214ab4cde/mealie/repos/repository_recipes.py#L153

**Comment by hay-kot** [comment]
> However, I also did some benchmarking with RapidFuzz (https://maxbachmann.github.io/RapidFuzz/) to test the crazy idea of just pulling relevant database fields from all recipes and then fuzzing against everything to figure out what matches. This would be database independent and could be easily pythonized inside the search function, but possibly slower. I was surprised to instead find out that RapidFuzz is blazingly fast and this is actually very fast with even a large dataset.

Probably want to include what hardware you're bench testing on. Lots of people run this of Pi's which I think will totally tank any feasibility here, but that's just a guess. Bet it runs great on my M1 though!

**Reply by jecorn** [reply]
Definitely a potential issue. But RapidFuzz uses a C++ backend for Levenshtein, which is the same as Postgres levenshtein() and sqlite editdist3(). Levenshtein doesn't need an index, so there's no db benefit there. Trigram is fast with an index, but there's still the complication of having two entirely differently modelled and syntaxed solutions for Postgres vs sqlite. So I (somewhat naively) think that the main overheads for RapidFuzz beyond having the database perform the search is pulling all records and creating the meta-string.

I should find a way to simulate a Pi for testing...

**Reply by fleshgolem** [reply]
You can reduce the amount of memory and cpu the app can use when you run it in a docker container. The last time I had an issue that seemed to be a problem on low spec hardware i set it to something like 512M RAM and half a core
https://docs.docker.com/config/containers/resource_constraints/

**Reply by jecorn** [reply]
And one can set this within the vscode docker container environment? Sorry for the naive question. This is my first time developing inside a container and I am a bit unclear on some of the stackexchange answers about passing docker arguments inside vscode.

**Comment by michael-genson** [comment]
This has since been implemented

**Comment by Tirkov** [comment]
<html>
<body>
<html><head></head><body><p><strong>Empirical test on v3.16.0 (Postgres backend) — multi-keyword search does AND, but fuzzy threshold too aggressive</strong></p>
<p>Following on from this discussion — sharing test results that suggest multi-keyword AND search is <em>partly</em> implemented on the Postgres backend, but with a fuzzy-matching threshold that produces false positives for short tokens. Posting in case it's useful for whoever picks this up next.</p>
<h3>Test setup</h3>
<ul>
<li>Mealie v3.16.0</li>
<li>Postgres 16 backend</li>
<li>4,653 recipes (Paprika import + manual additions)</li>
</ul>
<h3>API result counts</h3>
<p>Direct hits against <code>/api/recipes?search=...</code>:</p>

Query | Total
-- | --
(no search) | 4,653
chicken | 924
lemon | 497
chicken lemon | 387
"chicken lemon" (URL-encoded %22...%22) | 0
chicken AND lemon (literal "AND" as token) | 137


<p>The <code>chicken lemon</code> total (387) is <strong>smaller than both single-token totals</strong> — mathematically inconsistent with a pure OR search, so the backend is doing some form of AND/intersection.</p>
<h3>But fuzzy matching is too lenient on short tokens</h3>
<p>Sampling the first 10 results of <code>search=chicken+lemon</code>:</p>
<pre><code>✓ Lemon Chicken Scaloppine
✓ Herby Lemon Chicken with Greek Style Orzo Salad
✓ Succulent Lemon Chicken with Pesto Potato Salad
✓ Crispy Lemon Chicken With Chilli Rice
✓ Greek Lemon Chicken
✓ Crispy Lemon Chicken With Chilli Rice (1)
✓ Chicken and Lemon Orzo with fennel and Italian style cheese
✓ Lemon-Dill Chicken Sausage Couscous with Zucchini
✗ Midweek Chicken &amp; Leek Delight        ← no "lemon", "leek" matched fuzzy
✗ Chicken, Leek &amp; Potato Pie            ← no "lemon", "leek" matched fuzzy
</code></pre>
<p>8/10 are true matches. The other two contain "leek" — Postgres trigram similarity appears to be considering "leek" close enough to "lemon" (both 4-letter words, share <code>e</code>). Verified by inspecting recipe contents — neither has "lemon" in name, description, or ingredients.</p>
<h3>Suggested investigation</h3>
<ol>
<li>Tune the <code>pg_trgm</code> similarity threshold higher (currently the default ~0.3 is permissive for short tokens)</li>
<li>Or: skip fuzzy when token length ≤ 4 (most false positives are short words)</li>
<li>Or: provide a UI toggle / query syntax for "exact tokens" mode (something less restrictive than the current quoted-phrase behaviour, which requires exact word ordering)</li>
</ol>
<h3>Documentation gap</h3>
<p>Independent of the fuzzy issue: the docs describe "Smart search, mix &amp; match of 'quoted literal searches' and keyword search" without specifying that:</p>
<ul>
<li>Multi-word queries do AND (with fuzzy)</li>
<li>Quoted strings are exact-phrase-with-word-order, not token-level exactness</li>
<li>There's no boolean operator support (<code>AND</code>/<code>OR</code> as keywords are treated as literal tokens)</li>
</ul>
<p>Documenting current behaviour explicitly would prevent users (myself included) from migrating SQLite → Postgres expecting a different feature set.</p>
<p>Happy to help test if anyone picks this up.</p>
<p><em>Environment: v3.16.0, Postgres 16, 4,653 recipes, fresh migration from SQLite May 2026.</em></p></body></html>
</body>
</html>
